tracee/container_event_monitor_ebpf.c

```diff
@@ -476,6 +483,26 @@ static __always_inline u32 get_task_ppid(struct task_struct *task)
     return task->real_parent->pid;
 }
 
+static __always_inline void get_syscall_args(struct pt_regs *ctx, args_t *args)
+{
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+    args->args[0] = PT_REGS_PARM1(ctx);
+    args->args[1] = PT_REGS_PARM2(ctx);
+    args->args[2] = PT_REGS_PARM3(ctx);
+    args->args[3] = PT_REGS_PARM4(ctx);
+    args->args[4] = PT_REGS_PARM5(ctx);
+    args->args[5] = PT_REGS_PARM6(ctx);
+#else
+    struct pt_regs * ctx2 = (struct pt_regs *)ctx->di;
+    bpf_probe_read(&args->args[0], sizeof(args->args[0]), &ctx2->di);
+    bpf_probe_read(&args->args[1], sizeof(args->args[1]), &ctx2->si);
+    bpf_probe_read(&args->args[2], sizeof(args->args[2]), &ctx2->dx);
+    bpf_probe_read(&args->args[3], sizeof(args->args[3]), &ctx2->r10);
+    bpf_probe_read(&args->args[4], sizeof(args->args[4]), &ctx2->r8);
+    bpf_probe_read(&args->args[5], sizeof(args->args[5]), &ctx2->r9);
```

Comment:
Ah, so you're just manually reading from the registers? Does that macro provide anything besides that? If not, why bother checking kernel version?

Reply:
The macro does only this. The reason for the kernel version check is exactly because of the problem we saw between kernel versions. In kernels before 4.17, ctx has a different structure for systemcalls in x86 architecture - it was saved directly on ctx as pt_regs struct. From kernel 4.17, a new level of indirection was added, so now we have to first derefernce ctx->di, and only then we have pt_regs struct which can further dereference for getting the args

Reply:
Interesting! Thank you for the info!